File: Data Structure/20255/Solutions/Find_Parent_Value_at_Given_Index_in_Max_Heap.c
#include <stdio.h>
#include <stdlib.h>

int heap[1000], heapSize = 0;

void heapifyUp(int idx) {
    if (idx == 0) return;
    int parent = (idx - 1) / 2;
    if (heap[idx] > heap[parent]) {
        int temp = heap[idx];
        heap[idx] = heap[parent];
        heap[parent] = temp;
        heapifyUp(parent);
    }
}

void insert(int val) {
    heap[heapSize] = val;
    heapifyUp(heapSize);
    heapSize++;
}

int findParent(int index) {
    if (index == 0) return -1;
    if (index < 0 || index >= heapSize) return -1;
    int parentIdx = (index - 1) / 2;
    return heap[parentIdx];
}

int main() {
    int n; scanf("%d", &n);
    int arr[n];
    for (int i = 0; i < n; i++) scanf("%d", &arr[i]);
    int index; scanf("%d", &index);
    heapSize = 0;
    for (int i = 0; i < n; i++) insert(arr[i]);
    int result = findParent(index);
    printf("%d", result);
    return 0;
}

File: Data Structure/20255/Solutions/Find_Parent_Value_at_Given_Index_in_Max_Heap.java
import java.util.*;

class MaxHeap {
    private ArrayList<Integer> heap;
    
    public MaxHeap() {
        heap = new ArrayList<>();
    }
    
    public void insert(int val) {
        heap.add(val);
        heapifyUp(heap.size() - 1);
    }
    
    private void heapifyUp(int idx) {
        if (idx == 0) return;
        int parent = (idx - 1) / 2;
        if (heap.get(idx) > heap.get(parent)) {
            Collections.swap(heap, idx, parent);
            heapifyUp(parent);
        }
    }
    
    public int findParent(int index) {
        if (index == 0) return -1;
        if (index < 0 || index >= heap.size()) return -1;
        int parentIdx = (index - 1) / 2;
        return heap.get(parentIdx);
    }
    
    public ArrayList<Integer> getHeap() {
        return heap;
    }
}

class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int[] arr = new int[n];
        for (int i = 0; i < n; i++) arr[i] = sc.nextInt();
        int index = sc.nextInt();
        MaxHeap heap = new MaxHeap();
        for (int num : arr) heap.insert(num);
        int result = heap.findParent(index);
        System.out.println(result);
    }
}

